crates/pathfinder/src/cairo/ext_py/sub_process.rs

```diff
@@ -0,0 +1,303 @@
+            let _ = status_updates.send(SubProcessEvent::Failure(None, e)).await;
+            return None;
+        }
+    };
+
+    if status_updates
+        .send(SubProcessEvent::ProcessLaunched(pid))
+        .await
+        .is_err()
+    {
+        drop(input);
+        return None;
+    }
+
+    let mut command_buffer = Vec::new();
+
+    let exit_reason = loop {
+        let command = async {
+            let mut locked = commands.lock().await;
+            locked.recv().await
+        };
+
+        tokio::pin!(command);
+
+        let (call, at_block, response) = tokio::select! {
+            // locking is not cancellation safe BUT if the race is lost we don't retry so no
+            // worries on that.
+            maybe_command = &mut command => match maybe_command {
+                Some(tuple) => tuple,
+                None => break SubprocessExitReason::Shutdown,
+            },
+            _ = child.wait() => {
+                // if the python process was killed while we were awaiting for new commands, it
+                // would be zombie until we notice it has died. The wait can be called many times,
+                // and it'll return immediatedly at the top level.
+                break SubprocessExitReason::Death;
+            }
+            _ = shutdown_rx.recv() => {
+                break SubprocessExitReason::Shutdown;
+            },
+        };
+
+        command_buffer.clear();
+
+        let cmd = ChildCommand {
+            contract_address: &call.contract_address,
+            calldata: &call.calldata,
+            entry_point_selector: &call.entry_point_selector,
+            at_block: &at_block,
+        };
+
+        if let Err(e) = serde_json::to_writer(std::io::Cursor::new(&mut command_buffer), &cmd) {
+            eprintln!("Failed to render command {cmd:?} as json: {e:?}");
+            let _ = response.send(Err(CallFailure::Internal(
+                "Failed to render command as json",
+            )));
+            continue;
+        }
+
+        // using tokio::select to race against the shutdown_rx requires additional block to release
+        // the &mut borrow on buffer to have it printed/logged
+        let res = {
+            // AsyncWriteExt::write_all used in the rpc_round is not cancellation safe, but
+            // similar to above, if we lose the race, will kill the subprocess and get out.
+            let rpc_op = rpc_round(&command_buffer, &mut input, &mut output, &mut buffer);
+            tokio::pin!(rpc_op);
+
+            tokio::select! {
+                res = &mut rpc_op => res,
+                _ = shutdown_rx.recv() => {
+                    break SubprocessExitReason::Shutdown;
+                },
+                // no need to await for child dying here, because the event would close the childs
+                // stdout and thus break our read_line and thus return a SubprocessError::IO and
+                // we'd break out.
+            }
```

Comment:
There should be a timeout, maybe 60s?

Reply:
Maybe it doesn't make sense to impose arbitrary deadlines at this depth. We could get a notification is the user drops the oneshot.

Reply:
Done in https://github.com/eqlabs/pathfinder/pull/111/commits/4b18f9397a3b6fff5e7ece5d2413567e0487a83f